From 74f7764b20e6f519ee46b881d41b3f80a7762109 Mon Sep 17 00:00:00 2001
From: Greg Gallagher <greg@embeddedgreg.com>
Date: Tue, 6 Feb 2018 10:18:51 +0100
Subject: [PATCH 068/137] irqchip: dw-apb-ictl: ipipe: fix build

---
 drivers/irqchip/irq-dw-apb-ictl.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/irqchip/irq-dw-apb-ictl.c b/drivers/irqchip/irq-dw-apb-ictl.c
index 9b2696c..061189b 100644
--- a/drivers/irqchip/irq-dw-apb-ictl.c
+++ b/drivers/irqchip/irq-dw-apb-ictl.c
@@ -17,6 +17,7 @@
 #include <linux/irqchip/chained_irq.h>
 #include <linux/of_address.h>
 #include <linux/of_irq.h>
+#include <linux/ipipe.h>
 
 #define APB_INT_ENABLE_L	0x00
 #define APB_INT_ENABLE_H	0x04
-- 
1.9.1

